promql/engine.go

```diff
@@ -220,6 +234,9 @@ type EngineOpts struct {
 	// LookbackDelta determines the time since the last sample after which a time
 	// series is considered stale.
 	LookbackDelta time.Duration
+	SelectManager SelectManager
+	// MaxConcurrentSelect determines the maximum number of concurrent Selects per query.
```

Comment:
This isn't right, this isn't a PromQL concern.

This belongs in storage and in particular fanout.

Reply:
I disagree, we never use fanout anywhere beside Prometheus and we agreed we don't want concurrency for Prometheus now. So this have to live in PromQL. (: 

If that was so simple then we could already implement storage implementation the do things concurrently...  Can we do that without introducing any change to PromQL?


Also I don't think we want Select manager, some `concurrency` might work better...

Reply:
Now when I look at it I think we can totally do the concurrent implementation of querier which will work just fine.... :facepalm: 

So looks like concurrent select are done? Let's talk offline @kakkoyun If I am right then we can without any PromQL change do that in Thanos

Reply:
Pseudocode:

```
func (asynQueryable) Querier(....) storage.Querier {
  return new AsyncQuerier{} 
}

func (AsyncQuerier) Select(...) (set, w, err) {
  ret <- chan
  go func() {
     ret <- Select
  }
   return AsyncIterator{ret: ret}, nil, err
}

func (a AsyncIterator) Next() bool { 
   set <- a.resp
   // Iterate over set
}
```

With this we achieve full concurrency control. :heart: 

Is this what you thought about @brian-brazil ? The only downsides is lack of warning propagation... but that's not super critical I guess.

Reply:
A query that is only talking to the local tsdb should not spawn goroutines. PromQL should also not spawn gorountines. That sort of thing should be happening in fanout.

Remember we want to allow for concurrent remote read requests, but not anything else.

Reply:
@brian-brazil @bwplotka From the discussions here and IRC, I understood it is acceptable to change `storage.Querier` interface (which was the original implementation).
But I think I now understand what was wrong with it, as @brian-brazil has pointed out in https://github.com/prometheus/prometheus/pull/7198#discussion_r419337221.

I'm going to get rid of select manager and embed behaviour into Querier.  I'm going to try to split `.Select`s into a two-phased action. 

Also I would like to combine this approach with a lazy iterator/set pattern that @bwplotka suggested, to also get rid of the closure that wraps `.Select`s in PromQL engine.

To achieve that what do you think having an interface like below?

```diff
// Querier provides querying access over time-series data of a fixed time range.
type Querier interface {
	baseQuerier

	// Select returns a set of series that matches the given label matches.
	// For async .Select this would return a lazy iterator.
	// Can return scheduling errors, we can also keep Warnings.
-	Select(sortSeries bool, hints *SelectHints, matchers ...*labels.Matcher) (SeriesSet, Warnings, error)
+	Select(sortSeries bool, hints *SelectHints, matchers ...*labels.Matcher) (SeriesSet, error) 

+	// Exec returns aggregated warnings and errors occurred while executing scheduled Selects for a given Query.
+	Exec() (Warnings, error)
}
```

This implementation would require a minimum set of changes to TSDB implementation of Querier.
And also without touching underlying implementation `storage.Fanout` can wrap remote read querier for concurrent queries, if necessary.
 
What do you think?

Reply:
@bwplotka I know you have concerns about the interface change since it's commonly used. Proposed changes shouldn't create that much of a problem, do they?

Reply:
That interface seems like a good approach to me, we'll just need to make sure it's clear from the comments that you can't use the SeriesSet until Exec is called.

Reply:
Can we just have this interface change for PromQL? I don't we need this everywhere.

Reply:
I don't think we need anything else than just a tiny layer for sample series for promql purposes. Otherwise we will end up with tons of too wide interfaces for now reason.

Reply:
Example of this is `Seek` which ended up used only in one place and now all the deps have to have it for now reason and we are maintaining dead code.

Reply:
Let's keep interface as small as possible. Which means to me a separate interface in `promql` which can embed storage.Querier

Reply:
I'd rather not end up with two interfaces. Adding a call to allow things to be async isn't exactly making an interface wide.

Reply:
It makes because you will have noop in all the places like remote querier, block querier, and many others?

Reply:
Essentially what is expected behind such `Execute` method. It's not only `just one method` but thousands of small behaviour requirements.

Should it:
* Schedule concurrently?
* Schedule just one?
* Limit concurrency or consumer should limit?
* Retry?
* How to control concurrency? 
* Is retry with backoff?
* If one select failed in concurrent run inside of Execute should all be canceled or just one?

By adding `Execute` to interface IMO we end up with tons of unneeded complexity. That's why `ConcurrentQuerier` interface should only be one and possibly even one implementation based on normal `Querier.`

Reply:
The point of Execute is to allow the implementer to choose any of those behaviours, without having to hack it into Next. The complexity you mention is the feature we're trying to add here, forcing the implementation to be more complex on top of that is not a win.

I do not think we should have two subtly different Querier interfaces.

Reply:
But that's the point, you cannot hide all of this complexity under this interface. You have to know answers to most of those question if you are USING such interface with `Execute`. Otherwise how to ensure how concurrent we are? And what if I want one series Set samples like 1h earlier than second series Set? Why do I need to wait for both to finish? That's why I think even `Execute` might not enough ): 

Why `Next` is hacky? It's true streaming then. And if `Execute` and `Next` patterns are bad.. is there anything better? :thinking: 

And if we go with Execute I still think we should have separate interfaces as per rule to have it super simple. We do that all the time in other places. For example:

```
type ConcurrentQuerier interface {
     Querier

     Exec() (Warnings, error)
}
```
What's wrong with that?

Essentially we can make only engine requiring this so no need to implement this in all 3-4 queries we have just in Prometheus. (:

Reply:
I'm against the proliferation of interfaces, and would like all uses of Select to be consistent - otherwise that's just going to cause confusion down the line. Two interfaces seems like overengineering here, and with recent changes we already seem to have rather a lot of interfaces already.

Reply:
Actually we have the number of changes because of lack of fine-grained interfaces (: 

What about the downside I mentioned for Execute pattern? 

> And what if I want one series Set samples like 1h earlier than second series Set? Why do I need to wait for both to finish? That's why I think even Execute might not enough ):

Reply:
I didn't follow that example, but I guess in that cases you'd have two separate queriers as you're doing something very unusual.

Reply:
I don't see any difference between

```
Select(sortSeries bool, hints *SelectHints, matchers ...*labels.Matcher (SeriesSet, error) 
Exec() (Warnings, error)
```
vs making Select non blocking and handling the errors on `iterator.Next()`

The second option would even simplify the interface usage as you only handle errors once so I would vote for that one.

Reply:
The second one might have a simpler API, however the usage is quite intricate - as you can see in the current warnings handling.

Reply:
see where? this? 
https://github.com/prometheus/prometheus/pull/7198/files#diff-c180b6dbd8120d43794f4b192dc4a488R112-R119

Reply:
I mean the existing code, it was very tricky to get right - and that was the not the first attempt which had tried to do it via Next.

Reply:
I have been entertaining the ideas that we have discussed here: 

1. Splitting `storage.Querier`
https://github.com/prometheus/prometheus/compare/master...kakkoyun:concurrent_select_with_querier_changes
2. Add a new Interface `storage.ConcurrentQuerier`
https://github.com/prometheus/prometheus/compare/master...kakkoyun:concurrent_select_wrap_querier

Both have their own downsides. First one requires lots of changes. And from the consumer's perspective, it doesn't feel right. The second one is a slightly changed form of current PR. I believe it will the same concerns that have been voiced over here. (Of course, assuming this what you meant)

However, as far as I can tell if we can defer the error handling with the help of lazy/iterator.Next() that would be ideal. It's the simplest and requires no external API changes and it'd give a sufficient amount of flexibility. It requires changes in `promql/engine`.

What do you think, can we have a consensus on this?

@brian-brazil @bwplotka @krasi-georgiev

Reply:
> First one requires lots of changes. 

The goal is not to make the smallest PR possible, it's to have the overall simplest system to understand and maintain.

> And from the consumer's perspective, it doesn't feel right. 

API wise it seems fine to me, this is what changing a sync interface to an async interface looks like.

> I believe it will the same concerns that have been voiced over here.

Yes, it's putting concerns in places that they don't belong plus creating two subtly different querier interfaces which also then violates the Liskov substitution principle.

Reply:
What do you think about the alternative approach? Changing the error handling flow and deferring it until `iterator.Next()`

Reply:
That was tried before, it ended up really complicated. You can try it if you want, and it'll still not be okay for promql to have any knowledge of concurrency.

Reply:
>  it's to have the overall simplest system to understand and maintain.

:+1: 

Nice work, thanks @kakkoyun I think I agree with your thoughts.

Option 1. is really terrible. It has all the downsides I mentioned AND I can see even more now: It's easy to never run `exe` and the warnings, and errors, we don't know for which select..

Option 2 has the same issue, just we don't have this broken thing everywhere just in single place. Better than 1 =D

I think I fully agree with Kemal and the perfect thing would be to me:

Changing Querier interface to:
```
Select(sortSeries bool, hints *SelectHints, matchers ...*labels.Matcher) SeriesSet
```

And then changing SeriesSet to 
```
// SeriesSet contains a set of series.
type SeriesSet interface {
	Next() bool
	At() Series
	Err() error
        Warnings() storage.Warnings
}
```

This is IMO the most concise version, but it requires lot's of changes in the code for sure.

I think for now we can even start with implementing async select WITHOUT even any change in Prometheus, and slowly move with this interface change if we are happy with this.

We can do that because warnings does not matter much for now, plus we can workaround things in Thanos.

cc @brian-brazil @kakkoyun thoughts? (:

Reply:
I think, if it's acceptable I'm ok with this approach. Even so, I had already started implementing this. To understand how it goes. 
https://github.com/prometheus/prometheus/compare/master...kakkoyun:concurrent_select_iter_next

Reply:
I'm not seeing what's so terrible about option 1

> It's easy to never run exe

I'd expect a panic in that case, as that's a programming error.

> Option 2 has the same issue, just we don't have this broken thing everywhere just in single place. Better than 1 =D

I'm not at all convinced that having two subtly different ways of doing basically the same thing is a win. I'd much prefer a single consistent interface, otherwise we're tending towards an interface per invocation.

> We can do that because warnings does not matter much for now,

Warnings must maintain their current user-visible semantics from the HTTP APIs.

Reply:
> I think for now we can even start with implementing async select WITHOUT even any change in Prometheus, and slowly move with this interface change if we are happy with this.

As I said you can try the Next approach to see how viable it is, but it didn't work out last time and warnings need to keep on working.

Reply:
@brian-brazil @bwplotka @krasi-georgiev I've opened another one for the iterator solution. https://github.com/prometheus/prometheus/pull/7251. Please have a look.